spec/features/user_hpiopenid_spec.rb

```diff
@@ -34,12 +34,42 @@
     expect(user.uid).to eq(@uid)
   end
 
-  scenario "User logs in via HPI OpenID" do
+  scenario 'User logs in via HPI OpenID' do
     FactoryGirl.create(:user, email: @email, provider: @provider, uid: @uid)
 
     visit new_user_registration_path
     click_link I18n.t('devise.shared.links.sign_in_with_provider', :provider => 'HPI OpenID')
 
     expect(page).to have_css(".alert-success")
   end
+
+  scenario 'Account creation fails due to empty email' do
+    OmniAuth.config.mock_auth[:hpiopenid] = OmniAuth::AuthHash.new({
+                                                                       :provider => @provider,
+                                                                       :uid => @uid,
+                                                                       :info => {
+                                                                           :email => ''
+                                                                       }
+                                                                   })
+    visit new_user_registration_path
+    click_link I18n.t('devise.shared.links.sign_in_with_provider', :provider => 'HPI OpenID')
+
+    expect(page).to have_css(".alert-danger")
+    expect(page).to have_text (I18n.t('users.openid.missing_email'))
+  end
+
+  scenario 'Account creation fails due to missing email' do
```

Comment:
I am not sure how we should handle code duplication in tests. Currently the both tests are very similar.

Reply:
They are, but I don't know a better way. Would you create a method for the part that is the same? I think I'd find it weird to have the `expect` in a function and not the test itself. Unfortunately the big part in the beginning is the one that differs from test to test...

Reply:
I am not sure if these will make the code better, but I have some suggestions:

1. Loop around the scenarios:
``` ruby
['', nil].each {|invalid_email|
  scenario "Account creation fails if email equals #{invalid_email}" do
    OmniAuth.config.mock_auth[:hpiopenid] = OmniAuth::AuthHash.new({
                                                                       :provider => @provider,
                                                                       :uid => @uid,
                                                                       :info => {
                                                                           :email => invalid_email
                                                                       }
                                                                   })
    visit new_user_registration_path
    click_link I18n.t('devise.shared.links.sign_in_with_provider', :provider => 'HPI OpenID')

    expect(page).to have_css(".alert-danger")
    expect(page).to have_text (I18n.t('users.openid.missing_email'))
  end
}
```

2. Combine both into one scenario and add the loop there

3. Maybe parts of the scenario could be extracted into seperate methods, such as to create a valid AuthHash and change the value in the test to make them invalid?

Reply:
I have withdrawn my change request, because this is not a problem directly related to this PR, but a discussion in general how test should be lool like :)

Reply:
I really like the loop idea since this also extracts the nil values to a more visible place and makes the whole test more readable, not to mention it also solves the duplication problem here. I will implement your idea soon.